### PR TITLE
feat: add support for mysql and sqlite

### PR DESCRIPTION
**Pull request by Frank-III**
hopefully resolve: #67 

Hi there!
this is a simple hack that adds support for MySQL and sqlite, should be easy to implement for other dbs that sqlx has support for.
I haven't finished it yet, would like to discuss it before I continue working on it.

### CODE REVIEW

**Comment by achristmascarl** [comment]
hey @Frank-III, thanks for the PR and taking this on! i wasn't sure when i'd get to it, so this is really helpful 🙏 

structure-wise, I think everything makes sense so far. there's a few specific things i'll leave comments for, but overall it makes sense.

i'll also open a separate PR for mysql and sqlite containers/test data to make the development experience smoother, hopefully in time for you to make use of them lol

File: src/app.rs [hunk 1]
```diff
@@ -366,7 +375,8 @@ impl<'a> App<'a> {
             let query_string = query_lines.clone().join(" \n");
             if !query_string.is_empty() {
               self.add_to_history(query_lines.clone());
-              let first_query = database::get_first_query(query_string.clone());
+              let dialect = get_dialect(<DB as sqlx::Database>::NAME);
```
Comment (achristmascarl) [inline src/app.rs:378]: i think this should be moved to `main.rs`, run once, and stored as part of `app_state`

File: src/cli.rs [hunk 2]
```diff
@@ -28,5 +28,5 @@ pub struct Cli {
   pub port: Option<u16>,
 
   #[arg(long = "database", value_name = "DATABASE", help = "Name of database for connection (ex. postgres)")]
-  pub database: Option<String>,
+  pub database: String,
```
Comment (achristmascarl) [inline src/cli.rs:31]: this is actually referring to the name of the specific database within a postgres/mysql instance that is being connected to (not sure if sqlite has something similar), for example, postgres://root:password@localhost:5432/rainfrog, rainfrog is the "database" / "db_name"

for postgres vs mysql vs sqlite, we'll need a new arg, let's call it `driver`

File: src/main.rs [hunk 3]
```diff
@@ -35,10 +35,16 @@ async fn tokio_main() -> Result<()> {
   initialize_panic_handler()?;
 
   let args = Cli::parse();
-  let connection_opts = build_connection_opts(args.clone())?;
-  let mut app = App::new(connection_opts)?;
-  app.run().await?;
-
+  match args.database.as_str() {
```
Comment (achristmascarl) [inline src/main.rs:38]: `args.driver`

File: src/database/postgresql.rs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+      ("INSERT INTO users (name) VALUES ('John')", false),
+      ("EXPLAIN ANALYZE DELETE FROM users WHERE id = 1", true),
+      ("EXPLAIN SELECT * FROM users", false),
+      ("EXPLAIN ANALYZE SELECT * FROM users WHERE id = 1", false),
+    ];
+
+    for (query, expected) in test_cases {
+      let ast = Parser::parse_sql(&dialect, query).unwrap();
+      let statement = ast[0].clone();
+      // assert_eq!(should_use_tx(statement), expected, "Failed for query: {}", query);
```
Comment (achristmascarl) [inline src/database/postgresql.rs:288]: dild this test break?

File: src/components/data.rs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-use std::{collections::HashMap, sync::Arc, time::Duration};
+use std::{collections::HashMap, marker::PhantomData, sync::Arc, time::Duration};
```
Comment (achristmascarl) [inline src/components/data.rs:1]: looks like this import might have accidentally slipped in

File: src/components/editor.rs [hunk 6]
```diff
@@ -161,7 +174,7 @@ impl<'a> Component for Editor<'a> {
     Ok(None)
   }
 
-  fn update(&mut self, action: Action, app_state: &AppState) -> Result<Option<Action>> {
+  fn update(&mut self, action: Action, app_state: &AppState<'_, DB>) -> Result<Option<Action>> {
     match action {
```
Comment (achristmascarl) [inline src/components/editor.rs:178]: all these `MenuPreview` queries will need to be moved to their respective database drivers

**Comment by achristmascarl** [comment]
@Frank-III i just merged https://github.com/achristmascarl/rainfrog/pull/101, hopefully that will be helpful for you during development

**Comment by Frank-III** [comment]
> @Frank-III i just merged #101, hopefully that will be helpful for you during development

Thank you that would be very helpful, I would try to address all the issues and refer to the contribution guide!

**Comment by Frank-III** [comment]
Hi @achristmascarl, some updates on this PR.
The parser for MySQL and SQLite are almost working:
For MySQL: 
- Geometry type cannot be parsed correctly, we may need to parse it as binary and use crate like `geozero` to parse it and represent it as String.

For SQLite:
- types besides `INTEGER`, `REAL`, `TEXT`,  `BLOB` would be interpreted as `NULL` by `sqlx` and just cast as String for representation(like DECIMAL(10,2), JSON)
- We store geometry data as binary so casting it as binary and parsing it using `geozero` seems to be the solution as well.

Feel free to give it a try, and provide feedback! Thank you again for adding test data for MySQL and SQLite, super helpful!

File: src/database.rs [hunk 7]
```diff
@@ -34,64 +39,101 @@ pub struct Rows {
...
+  mut stream: BoxStream<'_, Result<Either<DB::QueryResult, DB::Row>, Error>>,
+) -> Result<Rows, DbError>
+where
+  DB: Database + ValueParser,
+  DB::QueryResult: HasRowsAffected,
+{
   let mut query_rows = vec![];
   let mut query_rows_affected: Option<u64> = None;
   let mut headers: Headers = vec![];
-  while !query_finished {
```
Comment (Frank-III) [inline src/database.rs:64]: I have to change this part as it hangs the MySQL connection. if I use 3 maximum connections, the third query will hang

**Comment by achristmascarl** [comment]
thanks @Frank-III, will take a look! we can hold off on the geometry types for now, i think i'll need to add either `geozero` or `georust` for #66 to support postgis, so we can wait until then and add it in for all the drivers together

**Comment by achristmascarl** [comment]
great work, the parsers and drivers set ups look good! two main points of feedback:

### 1. i don't think the mysql transactions work on drop statements
to repro:
1. init db, start app with mysql
2. `drop table robot_parts`
3. [N] to rollback
4. go to menu and [R] to refresh
5. robot_parts was dropped despite rolling back
6. however, rolling back seems to work for delete statements. if i try to delete all the rows and roll it back, the rows are not deleted
7. rolling back the drop statement in postgres still works

i'm not sure this is actually because of your code, it could be a quirk with sqlx and mysql. i'll do some more research tmrw as well

### 2. we should add some tests to the mysql and sqlite drivers, similar to what's there for postgres

## other
a minor suggestion, not blocking for the PR though, is to make the driver optional and get the user to input if it's missing

**Comment by achristmascarl** [comment]
it looks like drop statements can't be rolled back in mysql: https://dev.mysql.com/doc/refman/8.4/en/cannot-roll-back.html

i'll need to change the confirmation popup so that it can optionally show up before executing the query as well, you can skip this for now

**Comment by achristmascarl** [comment]
probably 'DROP' statements for all drivers can be "confirm before executing" instead of "execute in transaction"

**Comment by Frank-III** [comment]
> it looks like drop statements can't be rolled back in mysql: https://dev.mysql.com/doc/refman/8.4/en/cannot-roll-back.html
> probably 'DROP' statements for all drivers can be "confirm before executing" instead of "execute in transaction"

Yeah, that makes sense, I would skip this part for now then.

>2. we should add some tests to the mysql and sqlite drivers, similar to what's there for Postgres

I add the test for SQLite and MySQL, there is one SQLite query that could not get properly parsed by the `sqlparser` crate, and the `drop table` test for MySQL would fail when testing if we should use tx.

>a minor suggestion, not blocking for the PR though, is to make the driver optional and get the user to input if it's missing

I thought about this too! Also, if the user passes the URL directly we don't need the driver option as well

**Comment by achristmascarl** [comment]
thanks for adding in the driver implementation, and the tests look good! we can leave the other tests for now, i'll first release with a warning that mysql/sqlite are unstable and i'll revisit them later.

@Frank-III was there anything else you wanted to add or is the PR finished from your end? if it's finished, i can run the tests and add a few small things for Docker before merging :)

**Comment by Frank-III** [comment]
I think no, thank you for your prompt reply!

**Comment by achristmascarl** [comment]
great, i'm also going to have an automated review bot, but you can ignore it's feedback for now; it's just to surface things that i might need to take a second look at

@greptileai

**Pull request merged**